File: backend/app/schemas/shop_floor.py
from __future__ import annotations
from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ── Session ────────────────────────────────────────────────────────────────────

class SFSessionCreate(BaseModel):
    operator_name: str
    work_center_id: Optional[UUID] = None
    line_id: Optional[str] = None
    shift_id: Optional[UUID] = None
    machine_id: Optional[UUID] = None
    team_id: Optional[UUID] = None
    notes: Optional[str] = None


class SFSessionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    session_code: str
    operator_id: Optional[UUID]
    operator_name: str
    team_id: Optional[UUID]
    machine_id: Optional[UUID]
    line_id: Optional[str]
    work_center_id: Optional[UUID]
    shift_id: Optional[UUID]
    login_time: datetime
    logout_time: Optional[datetime]
    current_work_order_id: Optional[UUID]
    status: str
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime


# ── Activity Log ───────────────────────────────────────────────────────────────

class WOActivityLogOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    work_order_id: UUID
    production_order_id: Optional[UUID]
    session_id: Optional[UUID]
    operator_name: Optional[str]
    line_id: Optional[str]
    event_type: str
    timestamp_start: datetime
    timestamp_end: Optional[datetime]
    duration_minutes: Optional[Decimal]
    qty_good: Optional[Decimal]
    qty_scrap: Optional[Decimal]
    qty_rework: Optional[Decimal]
    reason_code: Optional[str]
    remarks: Optional[str]
    extra_data: Optional[dict]
    created_at: datetime


# ── Downtime ───────────────────────────────────────────────────────────────────

class SFDowntimeCreate(BaseModel):
    work_order_id: Optional[UUID] = None
    production_order_id: Optional[UUID] = None
    machine_id: Optional[UUID] = None
    machine_name: Optional[str] = None
    line_id: Optional[str] = None
    operator_name: Optional[str] = None
    downtime_category: str
    root_reason: Optional[str] = None
    impact_level: str = "MEDIUM"
    comments: Optional[str] = None


class SFDowntimeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    work_order_id: Optional[UUID]
    production_order_id: Optional[UUID]
    machine_id: Optional[UUID]
    machine_name: Optional[str]
    line_id: Optional[str]
    operator_name: Optional[str]
    start_time: datetime
    end_time: Optional[datetime]
    duration_minutes: Optional[Decimal]
    is_closed: bool
    downtime_category: str
    root_reason: Optional[str]
    impact_level: str
    comments: Optional[str]
    created_at: datetime


# ── Shift Handover ─────────────────────────────────────────────────────────────

class ShiftHandoverCreate(BaseModel):
    line_id: Optional[str] = None
    work_center_id: Optional[UUID] = None
    shift_from_id: Optional[UUID] = None
    shift_to_id: Optional[UUID] = None
    shift_from_name: Optional[str] = None
    shift_to_name: Optional[str] = None
    outgoing_supervisor: Optional[str] = None
    incoming_supervisor: Optional[str] = None
    outgoing_operators: Optional[List[str]] = None
    incoming_operators: Optional[List[str]] = None
    open_work_orders: Optional[List[dict]] = None
    qty_completed: Optional[Decimal] = None
    qty_remaining: Optional[Decimal] = None
    issues_open: Optional[List[str]] = None
    qc_pending_items: Optional[List[str]] = None
    material_shortages: Optional[List[str]] = None
    notes: Optional[str] = None


class ShiftHandoverOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    line_id: Optional[str]
    work_center_id: Optional[UUID]
    shift_from_name: Optional[str]
    shift_to_name: Optional[str]
    outgoing_supervisor: Optional[str]
    incoming_supervisor: Optional[str]
    outgoing_operators: Optional[list]
    incoming_operators: Optional[list]
    open_work_orders: Optional[list]
    qty_completed: Optional[Decimal]
    qty_remaining: Optional[Decimal]
    issues_open: Optional[list]
    qc_pending_items: Optional[list]
    material_shortages: Optional[list]
    notes: Optional[str]
    handover_time: datetime
    is_approved: bool
    approved_by: Optional[str]
    approved_at: Optional[datetime]
    created_at: datetime


# ── Supervisor Override ────────────────────────────────────────────────────────

class SupervisorOverrideCreate(BaseModel):
    work_order_id: Optional[UUID] = None
    production_order_id: Optional[UUID] = None
    supervisor_name: str
    override_type: str
    reason: Optional[str] = None
    target_operator_name: Optional[str] = None
    notes: Optional[str] = None


class SupervisorOverrideOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    work_order_id: Optional[UUID]
    production_order_id: Optional[UUID]
    supervisor_name: str
    override_type: str
    reason: Optional[str]
    target_operator_name: Optional[str]
    approved_at: datetime
    notes: Optional[str]
    created_at: datetime


# ── Action Requests ────────────────────────────────────────────────────────────

class StartWORequest(BaseModel):
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None
    line_id: Optional[str] = None
    machine_id: Optional[UUID] = None


class PauseWORequest(BaseModel):
    session_id: Optional[UUID] = None
    reason_code: Optional[str] = None
    remarks: Optional[str] = None
    operator_name: Optional[str] = None


class ResumeWORequest(BaseModel):
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None


class QtyUpdateRequest(BaseModel):
    qty_good: Decimal
    qty_scrap: Decimal = Decimal("0")
    qty_rework: Decimal = Decimal("0")
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None
    remarks: Optional[str] = None


class ScrapRequest(BaseModel):
    qty_scrap: Decimal
    reason_code: Optional[str] = None
    remarks: Optional[str] = None
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None


class FinishWORequest(BaseModel):
    qty_good: Decimal
    qty_scrap: Decimal = Decimal("0")
    qty_rework: Decimal = Decimal("0")
    run_time_actual: Optional[Decimal] = None
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None
    remarks: Optional[str] = None


class HelpRequest(BaseModel):
    remarks: str
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None


class QCTriggerRequest(BaseModel):
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None
    remarks: Optional[str] = None


class NoteRequest(BaseModel):
    remarks: str
    session_id: Optional[UUID] = None
    operator_name: Optional[str] = None


class StopDowntimeRequest(BaseModel):
    comments: Optional[str] = None


# ── Live Queue & Status ────────────────────────────────────────────────────────

class LiveWOItem(BaseModel):
    id: str
    work_order_id: str
    production_order_id: Optional[str]
    order_no: Optional[str]
    step_sequence: int
    step_name: str
    work_center_id: Optional[str]
    line_id: Optional[str]
    planned_qty: float
    completed_qty: float
    scrap_qty: float
    status: str               # from ExecWorkOrder
    live_status: str          # derived shop floor status
    started_at: Optional[str]
    planned_start: Optional[str]
    planned_end: Optional[str]
    product_name: Optional[str]
    batch_no: Optional[str]
    qc_required_flag: bool
    has_active_downtime: bool
    last_event: Optional[str]
    last_event_time: Optional[str]
    operator_name: Optional[str]


class LiveStatusResponse(BaseModel):
    total_work_orders: int
    running: int
    paused: int
    waiting: int
    qc_hold: int
    active_sessions: int
    active_downtime: int
    alerts: List[dict]
    by_line: Dict[str, Any]
    timestamp: str


# ── AI ─────────────────────────────────────────────────────────────────────────

class SFAIRecOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    work_order_id: Optional[UUID]
    production_order_id: Optional[UUID]
    agent_type: str
    status: str
    title: str
    explanation: Optional[str]
    impact_summary: Optional[str]
    confidence_score: Optional[Decimal]
    actioned_at: Optional[datetime]
    created_at: datetime


# ── Dashboard ──────────────────────────────────────────────────────────────────

class SFDashboard(BaseModel):
    active_sessions: int
    running_work_orders: int
    paused_work_orders: int
    qc_hold_count: int
    active_downtime_count: int
    help_requests_open: int
    total_scrap_today: float
    total_good_qty_today: float
    pending_ai_recs: int
    recent_handovers: List[ShiftHandoverOut]
    active_downtimes: List[SFDowntimeOut]
    supervisor_overrides_today: int

File: frontend/src/lib/shopFloor.ts
const BASE = "/api/v1/shop-floor";

// ── Types ──────────────────────────────────────────────────────────────────────

export type SFSessionStatus = "ACTIVE" | "IDLE" | "LOGGED_OUT" | "EXPIRED";

export type WOEventType =
  | "START" | "PAUSE" | "RESUME" | "FINISH"
  | "QTY_UPDATE" | "SCRAP" | "DOWNTIME_START" | "DOWNTIME_STOP"
  | "QC_TRIGGER" | "HELP_REQUEST" | "HANDOVER" | "NOTE" | "SUPERVISOR_OVERRIDE";

export type SFLiveStatus =
  | "WAITING" | "READY" | "ASSIGNED" | "RUNNING"
  | "PAUSED" | "BLOCKED" | "QC_HOLD" | "SUPERVISOR_REVIEW"
  | "COMPLETED" | "CANCELLED";

export type SFDowntimeCat =
  | "MACHINE_BREAKDOWN" | "MATERIAL_SHORTAGE" | "PACKAGING_SHORTAGE"
  | "QC_WAITING" | "CHANGEOVER" | "CLEANING" | "POWER_ISSUE"
  | "LABOR_SHORTAGE" | "SUPERVISOR_HOLD" | "MAINTENANCE_STOP"
  | "LABEL_MISMATCH" | "IT_ISSUE" | "OTHER";

export type SFImpactLevel = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";

export interface SFSession {
  id: string;
  session_code: string;
  operator_id: string | null;
  operator_name: string;
  team_id: string | null;
  machine_id: string | null;
  line_id: string | null;
  work_center_id: string | null;
  shift_id: string | null;
  login_time: string;
  logout_time: string | null;
  current_work_order_id: string | null;
  status: SFSessionStatus;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface WOActivityLog {
  id: string;
  work_order_id: string;
  production_order_id: string | null;
  session_id: string | null;
  operator_name: string | null;
  line_id: string | null;
  event_type: WOEventType;
  timestamp_start: string;
  timestamp_end: string | null;
  duration_minutes: number | null;
  qty_good: number | null;
  qty_scrap: number | null;
  qty_rework: number | null;
  reason_code: string | null;
  remarks: string | null;
  extra_data: Record<string, unknown> | null;
  created_at: string;
}

export interface SFDowntimeLog {
  id: string;
  work_order_id: string | null;
  production_order_id: string | null;
  machine_id: string | null;
  machine_name: string | null;
  line_id: string | null;
  operator_name: string | null;
  start_time: string;
  end_time: string | null;
  duration_minutes: number | null;
  is_closed: boolean;
  downtime_category: SFDowntimeCat;
  root_reason: string | null;
  impact_level: SFImpactLevel;
  comments: string | null;
  created_at: string;
}

export interface ShiftHandover {
  id: string;
  line_id: string | null;
  work_center_id: string | null;
  shift_from_name: string | null;
  shift_to_name: string | null;
  outgoing_supervisor: string | null;
  incoming_supervisor: string | null;
  outgoing_operators: string[] | null;
  incoming_operators: string[] | null;
  open_work_orders: Record<string, unknown>[] | null;
  qty_completed: number | null;
  qty_remaining: number | null;
  issues_open: string[] | null;
  qc_pending_items: string[] | null;
  material_shortages: string[] | null;
  notes: string | null;
  handover_time: string;
  is_approved: boolean;
  approved_by: string | null;
  approved_at: string | null;
  created_at: string;
}

export interface SupervisorOverride {
  id: string;
  work_order_id: string | null;
  production_order_id: string | null;
  supervisor_name: string;
  override_type: string;
  reason: string | null;
  target_operator_name: string | null;
  approved_at: string;
  notes: string | null;
  created_at: string;
}

export interface LiveWOItem {
  id: string;
  work_order_id: string;
  production_order_id: string | null;
  order_no: string | null;
  step_sequence: number;
  step_name: string;
  work_center_id: string | null;
  line_id: string | null;
  planned_qty: number;
  completed_qty: number;
  scrap_qty: number;
  status: string;
  live_status: SFLiveStatus;
  started_at: string | null;
  planned_start: string | null;
  planned_end: string | null;
  product_name: string | null;
  batch_no: string | null;
  qc_required_flag: boolean;
  has_active_downtime: boolean;
  last_event: string | null;
  last_event_time: string | null;
  operator_name: string | null;
}

export interface LiveStatusResponse {
  total_work_orders: number;
  running: number;
  paused: number;
  waiting: number;
  qc_hold: number;
  active_sessions: number;
  active_downtime: number;
  alerts: Record<string, unknown>[];
  by_line: Record<string, unknown>;
  timestamp: string;
}

export interface SFDashboard {
  active_sessions: number;
  running_work_orders: number;
  paused_work_orders: number;
  qc_hold_count: number;
  active_downtime_count: number;
  help_requests_open: number;
  total_scrap_today: number;
  total_good_qty_today: number;
  pending_ai_recs: number;
  recent_handovers: ShiftHandover[];
  active_downtimes: SFDowntimeLog[];
  supervisor_overrides_today: number;
}

export interface SFAIRec {
  id: string;
  work_order_id: string | null;
  production_order_id: string | null;
  agent_type: string;
  status: string;
  title: string;
  explanation: string | null;
  impact_summary: string | null;
  confidence_score: number | null;
  actioned_at: string | null;
  created_at: string;
}

// ── Color maps ─────────────────────────────────────────────────────────────────

export const LIVE_STATUS_COLOR: Record<SFLiveStatus, string> = {
  WAITING:           "bg-gray-100 text-gray-500",
  READY:             "bg-blue-100 text-blue-700",
  ASSIGNED:          "bg-indigo-100 text-indigo-700",
  RUNNING:           "bg-green-100 text-green-700",
  PAUSED:            "bg-yellow-100 text-yellow-700",
  BLOCKED:           "bg-red-100 text-red-700",
  QC_HOLD:           "bg-purple-100 text-purple-700",
  SUPERVISOR_REVIEW: "bg-orange-100 text-orange-700",
  COMPLETED:         "bg-teal-100 text-teal-700",
  CANCELLED:         "bg-gray-200 text-gray-400",
};

export const DOWNTIME_CAT_LABEL: Record<SFDowntimeCat, string> = {
  MACHINE_BREAKDOWN:  "Machine Breakdown",
  MATERIAL_SHORTAGE:  "Material Shortage",
  PACKAGING_SHORTAGE: "Packaging Shortage",
  QC_WAITING:         "QC Waiting",
  CHANGEOVER:         "Changeover",
  CLEANING:           "Cleaning",
  POWER_ISSUE:        "Power Issue",
  LABOR_SHORTAGE:     "Labor Shortage",
  SUPERVISOR_HOLD:    "Supervisor Hold",
  MAINTENANCE_STOP:   "Maintenance Stop",
  LABEL_MISMATCH:     "Label Mismatch",
  IT_ISSUE:           "IT Issue",
  OTHER:              "Other",
};

export const IMPACT_COLOR: Record<SFImpactLevel, string> = {
  LOW:      "bg-green-100 text-green-700",
  MEDIUM:   "bg-yellow-100 text-yellow-700",
  HIGH:     "bg-orange-100 text-orange-700",
  CRITICAL: "bg-red-100 text-red-700",
};

// ── API client ─────────────────────────────────────────────────────────────────

async function req<T>(path: string, opts?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...opts,
  });
  if (!res.ok) throw new Error(`${res.status} ${await res.text()}`);
  return res.json();
}

export const sfApi = {
  // Dashboard & status
  getDashboard: () => req<SFDashboard>("/dashboard"),
  getLiveStatus: () => req<LiveStatusResponse>("/live-status"),
  getLiveQueue: (params?: { line_id?: string }) => {
    const qs = params?.line_id ? `?line_id=${params.line_id}` : "";
    return req<LiveWOItem[]>(`/live-queue${qs}`);
  },
  getAlerts: () => req<Record<string, unknown>[]>("/alerts"),

  // Sessions
  createSession: (body: Partial<SFSession> & { operator_name: string }) =>
    req<SFSession>("/sessions", { method: "POST", body: JSON.stringify(body) }),
  listSessions: (status?: string) => {
    const qs = status ? `?status=${status}` : "";
    return req<SFSession[]>(`/sessions${qs}`);
  },
  getActiveSessions: () => req<SFSession[]>("/sessions/active"),
  logoutSession: (id: string) =>
    req<SFSession>(`/sessions/${id}/logout`, { method: "POST", body: "{}" }),

  // WO execution
  startWO: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/start`, { method: "POST", body: JSON.stringify(body) }),
  pauseWO: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/pause`, { method: "POST", body: JSON.stringify(body) }),
  resumeWO: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/resume`, { method: "POST", body: JSON.stringify(body) }),
  updateQty: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/update-qty`, { method: "POST", body: JSON.stringify(body) }),
  recordScrap: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/scrap`, { method: "POST", body: JSON.stringify(body) }),
  finishWO: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/finish`, { method: "POST", body: JSON.stringify(body) }),
  requestHelp: (woId: string, remarks: string, operatorName?: string) =>
    req<WOActivityLog>(`/work-orders/${woId}/help`, { method: "POST", body: JSON.stringify({ remarks, operator_name: operatorName }) }),
  triggerQC: (woId: string, body: Record<string, unknown>) =>
    req<WOActivityLog>(`/work-orders/${woId}/qc-trigger`, { method: "POST", body: JSON.stringify(body) }),
  addNote: (woId: string, remarks: string, operatorName?: string) =>
    req<WOActivityLog>(`/work-orders/${woId}/note`, { method: "POST", body: JSON.stringify({ remarks, operator_name: operatorName }) }),
  getActivityLog: (woId: string) =>
    req<WOActivityLog[]>(`/work-orders/${woId}/activity`),

  // Downtime
  startDowntime: (body: Record<string, unknown>) =>
    req<SFDowntimeLog>("/downtime", { method: "POST", body: JSON.stringify(body) }),
  listDowntime: (params?: { is_closed?: boolean; line_id?: string }) => {
    const qs = new URLSearchParams();
    if (params?.is_closed !== undefined) qs.set("is_closed", String(params.is_closed));
    if (params?.line_id) qs.set("line_id", params.line_id);
    return req<SFDowntimeLog[]>(`/downtime${qs.toString() ? "?" + qs : ""}`);
  },
  stopDowntime: (id: string, comments?: string) =>
    req<SFDowntimeLog>(`/downtime/${id}/stop`, { method: "POST", body: JSON.stringify({ comments }) }),

  // Handovers
  createHandover: (body: Record<string, unknown>) =>
    req<ShiftHandover>("/handovers", { method: "POST", body: JSON.stringify(body) }),
  listHandovers: (lineId?: string) => {
    const qs = lineId ? `?line_id=${lineId}` : "";
    return req<ShiftHandover[]>(`/handovers${qs}`);
  },
  approveHandover: (id: string, approvedBy: string) =>
    req<ShiftHandover>(`/handovers/${id}/approve`, { method: "POST", body: JSON.stringify({ approved_by: approvedBy }) }),

  // Overrides
  createOverride: (body: Record<string, unknown>) =>
    req<SupervisorOverride>("/overrides", { method: "POST", body: JSON.stringify(body) }),
  listOverrides: () => req<SupervisorOverride[]>("/overrides"),

  // AI
  runAI: () => req<Record<string, number>>("/ai/run", { method: "POST", body: "{}" }),
  listAIRecs: (params?: { status?: string; agent_type?: string }) => {
    const qs = new URLSearchParams();
    if (params?.status) qs.set("status", params.status);
    if (params?.agent_type) qs.set("agent_type", params.agent_type);
    return req<SFAIRec[]>(`/ai/recs${qs.toString() ? "?" + qs : ""}`);
  },
  actionAIRec: (id: string, status: "ACCEPTED" | "REJECTED") =>
    req<SFAIRec>(`/ai/recs/${id}/action`, { method: "POST", body: JSON.stringify({ status }) }),
};
